### ISSUE TITLE
Unable to get uncompressed ram image from buffer with multisamples enabled.

### ISSUE THREAD

**Issue by Clockwork-Muse** [panda3d/panda3d#1129]
## Description
When rendering to an FBO with multisamples enabled, the uncompressed ram image is black/transparent.

In my earlier issues I was concentrating too much on removing dependencies, and didn't catch where things were really happening.  This better reflects my current anticipated usage, as my project will need the uncompressed ram image for final processing on the CPU before persisting it.

## Steps to Reproduce
This script needs opencv+numpy to translate the uncompressed ram image to a more normal format, so first `pip install numpy opencv-python-headless`

```python
from direct.showbase.ShowBase import ShowBase
from panda3d.core import AmbientLight, FrameBufferProperties, GraphicsOutput, GraphicsPipe, LColor, LVector2i, Texture, WindowProperties, loadPrcFileData
import numpy
import cv2

class MyApp(ShowBase):
    def __init__(self):
        super().__init__(self)

        # Explicitly disable default camera.
        self.cam.node().set_active(False)

        window_props = WindowProperties(size=LVector2i(1024, 768))
        flags = GraphicsPipe.BFRefuseWindow

        color_props = FrameBufferProperties()
        color_props.set_rgba_bits(8, 8, 8, 0)
        color_props.set_float_color(False)
        color_props.set_depth_bits(16)
        color_props.set_float_depth(True)
        # This line causes issues
        color_props.set_multisamples(4)

        self.color_texture = Texture()
        self.color_buffer = self.graphicsEngine.make_output(self.pipe, "color buffer", -1, color_props, window_props, flags, self.win.get_gsg(), self.win)
        self.color_buffer.add_render_texture(tex=self.color_texture, mode=GraphicsOutput.RTMCopyRam, bitplane=GraphicsOutput.RTPColor)
        self.color_buffer.get_texture().set_clear_color(LColor(0, 0, 0, 0))
        color_camera = self.make_camera(self.color_buffer, camName="color")
        color_camera.reparent_to(self.render)
        color_camera.set_pos(7, 7, 7)
        color_camera.look_at(0, 0, 0)

        model = self.loader.loadModel("untitled.blend")
        model.reparent_to(self.render)

        light_node = AmbientLight('light')
        light_node.set_color((0.8, 0.8, 0.5, 1))
        light = self.render.attach_new_node(light_node)
        self.render.set_light(light)


def generate_synth():
    loadPrcFileData("",
    """
        win-size 1024 768
        window-type offscreen
        model-cache-dir
    """)

    base = MyApp()
    base.graphicsEngine.render_frame()
    base.graphicsEngine.render_frame()
    base.graphicsEngine.render_frame()
    base.graphicsEngine.render_frame()

    texture = base.color_texture
    ram_image = texture.get_uncompressed_ram_image()

    numpy_buffer = numpy.empty((40, 40, 3), dtype=numpy.uint8)
    buffer_view = numpy.frombuffer(ram_image, dtype=numpy_buffer.dtype)
    new_shape = ((texture.y_size, texture.x_size, 3))
    buffer_view.shape = new_shape

    numpy_buffer.resize(new_shape, refcheck=False)
    numpy.copyto(numpy_buffer, buffer_view)


    cv2.imwrite(f"screenshot.jpg", numpy_buffer)

generate_synth()
```

## Workarounds

Using this buffer as the initialization window for simplepbr outputs a working image:
```python
# Have to set buffer multisamples off, though
color_props.set_multisamples(0)

simplepbr.init(
    use_normal_maps=True,
    enable_shadows=True,
    use_occlusion_maps=True,
    window=self.__color_buffer,
    camera_node=color_camera,
    msaa_samples=4,
)
```

... which I probably need to be doing anyways, to make sure that color and lighting effects match.

## Environment
 - Operating system: Ubuntu 20.04
 - Docker Container base: nvidia/opengl:1.2-glvnd-runtime-ubuntu20.04
 - System architecture: x86_64
 - Panda3D version: 1.10.8; 1.11.0
 - Installation method: PyPI package for 10.8; built from source, commit ff80adc68046218f1523e1b659c9a9a176e26fd0 for 1.11
 - Python version: 3.8.5

**Comment by rdb** [comment]
I get this error when I run it (with GLX, not EGL):
```
:display:gsg:glgsg(error): GL_INVALID_OPERATION in glReadPixels(multisample FBO)
```
It looks like we are trying to read from the multisample FBO instead of from the resolved FBO.  This means that the transfer failed entirely.

Likely a similar fix as 5cadd86d1edf86182347d4d14b6f95cf108aa57c.  In fact I'm wondering why that commit didn't fix this.

**Comment by rdb** [comment]
This is fixed, but I have noticed there are still issues with multisample FBOs when using copy-texture or copy-ram with multiple texture pages (such as cube maps).  That is harder to fix.  A workaround is to use copy-ram mode and `engine.extract_texture_data(tex)`.
